tests/common/dualtor/server_traffic_utils.py

```diff
@@ -0,0 +1,90 @@
+"""Utils to verify traffic between ToR and server."""
+import contextlib
+import logging
+import tempfile
+import sys
+import time
+
+from io import BytesIO
+from ptf.dataplane import match_exp_pkt
+from scapy.all import sniff
+from scapy.packet import ls
+
+
+@contextlib.contextmanager
+def dump_intf_packets(ansible_host, iface, pcap_save_path, dumped_packets, pcap_filter=None):
+    """Dump packets of the interface and save to a file."""
+    start_pcap = "tcpdump --immediate-mode -i %s -w %s" % (iface, pcap_save_path)
+    if pcap_filter:
+        start_pcap += (" " + pcap_filter)
+    start_pcap = "nohup %s  > /dev/null 2>&1 & echo $!" % start_pcap
+    pid = ansible_host.shell(start_pcap)["stdout"]
+    # sleep to let tcpdump starts to capture
+    time.sleep(1)
+    try:
+        yield
+    finally:
+        ansible_host.shell("kill -s 2 %s" % pid)
+        with tempfile.NamedTemporaryFile() as temp_pcap:
+            ansible_host.fetch(src=pcap_save_path, dest=temp_pcap.name, flat=True)
+            packets = sniff(offline=temp_pcap.name)
+            dumped_packets.extend(packets)
+
+
+class ServerTrafficMonitor(object):
+    """Monit traffic between ToR and server."""
+
+    VLAN_INTERFACE_TEMPLATE = "{external_port}.{vlan_id}"
+
+    def __init__(self, duthost, vmhost, dut_iface, conn_graph_facts, exp_pkt, existing=True):
+        self.duthost = duthost
+        self.dut_iface = dut_iface
+        self.exp_pkt = exp_pkt
+        self.vmhost = vmhost
+        self.conn_graph_facts = conn_graph_facts
+        self.captured_packets = []
+        self.matched_packets = []
+        self.vmhost_iface = self._find_vmhost_vlan_interface()
+        self.dump_utility = dump_intf_packets(
+            vmhost,
+            self.vmhost_iface,
+            tempfile.NamedTemporaryFile().name,
+            self.captured_packets
+        )
+        self.existing = existing
+
+    @staticmethod
+    def _list_layer_str(packet):
+        """Return list layer output string."""
+        _stdout, sys.stdout = sys.stdout, BytesIO()
+        try:
+            ls(packet)
+            return sys.stdout.getvalue()
+        finally:
+            sys.stdout = _stdout
+
+    def _find_vmhost_vlan_interface(self):
+        """Find the vmhost vlan interface that will be sniffed on."""
+        device_port_vlans = self.conn_graph_facts["device_port_vlans"][self.duthost.hostname]
+        vlan_id = device_port_vlans[self.dut_iface]["vlanlist"][0]
+        return self.VLAN_INTERFACE_TEMPLATE.format(external_port=self.vmhost.external_port, vlan_id=vlan_id)
+
+    def __enter__(self):
+        self.captured_packets[:] = []
+        self.matched_packets[:] = []
+        self.dump_utility.__enter__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
```

Comment:
Is the ServerTrafficMonitor instance destroyed upon exit of context? Then the captured packets is not accessible after that. Is there a way to keep the captured packets for further analysis?

Reply:
No, the packets are always accessible from `captured_packets` attribute of the `ServerTrafficMonitor` instance.

Reply:
I see, we can use code like below to keep the `ServerTrafficMonitor` instance:
```
with ServerTrafficMonitor(...) as stm:
    # do stuff
# stm is still accessible here.
```